OpenRA.Mods.Common/Traits/Player/PlayerRadarTerrain.cs

```diff
@@ -0,0 +1,115 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using OpenRA.Graphics;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class PlayerRadarTerrainInfo : ITraitInfo
+	{
+		public object Create(ActorInitializer init)
+		{
+			return new PlayerRadarTerrain(init.Self);
+		}
+	}
+
+	public class PlayerRadarTerrain : IWorldLoaded
+	{
+		readonly World world;
+		CellLayer<Pair<int, int>> terrainColor;
+		readonly HashSet<MPos> dirtyTerrainCells = new HashSet<MPos>();
+		readonly Shroud shroud;
+
+		public event Action<MPos> CellTerrainColorChanged = null;
+
+		public PlayerRadarTerrain(Actor self)
+		{
+			world = self.World;
+			shroud = self.Trait<Shroud>();
+			shroud.CellsChanged += ShroudOnCellsChanged;
+		}
+
+		void ShroudOnCellsChanged(IEnumerable<PPos> puvs)
+		{
+			foreach (var puv in puvs)
+			{
+				foreach (var uv in world.Map.Unproject(puv))
```

Comment:
This may be a dumb question, but I wanted to get a quick review off before I disappear: why are we using `Unproject` etc here when I don't see a similar removal from RadarWidget`?

Reply:
Because we still handle the Shroud in the RadarWidget (for drawing the shroud) so this method is based on `RadarWidget.UpdateShroudCell()`